### ISSUE TITLE
Crash when all playlist entries are deleted

### ISSUE THREAD

**Issue by Digitalone1** [zaps166/QMPlay2#796]
I opened a folder with various subfolders containing some videos and they got all listed in the playlist sidebar.

Unfortunately, when I did CTRL+A to select them all and CANC to clear the playlist, the app crashed. This is easily reproducible when the playlist is populated with several entries and lots of subfolders.

Launching from the command line and reproducing it another time, I see a segmentation fault.

I managed to catch a core dump:

<details>
<summary> Stack trace </summary>

```
                Stack trace of thread 70951:
                #0  0x00005665c617a4b5 n/a (QMPlay2 + 0x7a4b5)
                #1  0x00005665c617ca0f n/a (QMPlay2 + 0x7ca0f)
                #2  0x000077da0cfb1b29 n/a (libQt6Core.so.6 + 0x1b1b29)
                #3  0x000077da0d967484 _ZN7QAction8activateENS_11ActionEventE (libQt6Gui.so.6 + 0x567484)
                #4  0x000077da0d969caa _ZN7QAction5eventEP6QEvent (libQt6Gui.so.6 + 0x569caa)
                #5  0x000077da0defe78a _ZN19QApplicationPrivate13notify_helperEP7QObjectP6QEvent (libQt6Widgets.so.6 + 0xfe78a)
                #6  0x000077da0cf57ac8 _ZN16QCoreApplication15notifyInternal2EP7QObjectP6QEvent (libQt6Core.so.6 + 0x157ac8)
                #7  0x000077da0d98230a _ZN12QShortcutMap13dispatchEventEP9QKeyEvent (libQt6Gui.so.6 + 0x58230a)
                #8  0x000077da0d5f8a70 _ZN22QWindowSystemInterface19handleShortcutEventEP7QWindowmi6QFlagsIN2Qt16KeyboardModifierEEjjjRK7QStringbt (libQt6Gui.so.6 + 0x1f8a70)
                #9  0x000077da0d588b8a _ZN22QGuiApplicationPrivate15processKeyEventEPN29QWindowSystemInterfacePrivate8KeyEventE (libQt6Gui.so.6 + 0x188b8a)
                #10 0x000077da0d602b44 _ZN22QWindowSystemInterface22sendWindowSystemEventsE6QFlagsIN10QEventLoop17ProcessEventsFlagEE (libQt6Gui.so.6 + 0x202b44)
                #11 0x000077da0da4ea14 n/a (libQt6Gui.so.6 + 0x64ea14)
                #12 0x000077da0a46a559 n/a (libglib-2.0.so.0 + 0x5d559)
                #13 0x000077da0a4cd157 n/a (libglib-2.0.so.0 + 0xc0157)
                #14 0x000077da0a469a55 g_main_context_iteration (libglib-2.0.so.0 + 0x5ca55)
                #15 0x000077da0d1c22fd _ZN20QEventDispatcherGlib13processEventsE6QFlagsIN10QEventLoop17ProcessEventsFlagEE (libQt6Core.so.6 + 0x3c22fd)
                #16 0x000077da0cf61ee6 _ZN10QEventLoop4execE6QFlagsINS_17ProcessEventsFlagEE (libQt6Core.so.6 + 0x161ee6)
                #17 0x000077da0cf58f4f _ZN16QCoreApplication4execEv (libQt6Core.so.6 + 0x158f4f)
                #18 0x00005665c6142d96 n/a (QMPlay2 + 0x42d96)
                #19 0x000077da0c834e08 n/a (libc.so.6 + 0x25e08)
                #20 0x000077da0c834ecc __libc_start_main (libc.so.6 + 0x25ecc)
                #21 0x00005665c6135255 n/a (QMPlay2 + 0x35255)
                
                Stack trace of thread 70952:
                #0  0x000077da0c91abb0 ppoll (libc.so.6 + 0x10bbb0)
                #1  0x000077da0a4cd227 n/a (libglib-2.0.so.0 + 0xc0227)
                #2  0x000077da0a469a55 g_main_context_iteration (libglib-2.0.so.0 + 0x5ca55)
                #3  0x000077da0d1c22fd _ZN20QEventDispatcherGlib13processEventsE6QFlagsIN10QEventLoop17ProcessEventsFlagEE (libQt6Core.so.6 + 0x3c22fd)
                #4  0x000077da0cf61ee6 _ZN10QEventLoop4execE6QFlagsINS_17ProcessEventsFlagEE (libQt6Core.so.6 + 0x161ee6)
                #5  0x000077da0d057f22 _ZN7QThread4execEv (libQt6Core.so.6 + 0x257f22)
                #6  0x000077da0dd735ae n/a (libQt6DBus.so.6 + 0x2e5ae)
                #7  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #8  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #9  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70953:
                #0  0x000077da0c9261fd syscall (libc.so.6 + 0x1171fd)
                #1  0x000077da0a49bf20 g_cond_wait (libglib-2.0.so.0 + 0x8ef20)
                #2  0x000077da0a43298c n/a (libglib-2.0.so.0 + 0x2598c)
                #3  0x000077da0a4a2137 n/a (libglib-2.0.so.0 + 0x95137)
                #4  0x000077da0a49e026 n/a (libglib-2.0.so.0 + 0x91026)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70956:
                #0  0x000077da0c91abb0 ppoll (libc.so.6 + 0x10bbb0)
                #1  0x000077da0a4cd227 n/a (libglib-2.0.so.0 + 0xc0227)
                #2  0x000077da0a46b287 g_main_loop_run (libglib-2.0.so.0 + 0x5e287)
                #3  0x000077d9f9f44bc4 n/a (libgio-2.0.so.0 + 0x113bc4)
                #4  0x000077da0a49e026 n/a (libglib-2.0.so.0 + 0x91026)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70954:
                #0  0x000077da0c91abb0 ppoll (libc.so.6 + 0x10bbb0)
                #1  0x000077da0a4cd227 n/a (libglib-2.0.so.0 + 0xc0227)
                #2  0x000077da0a469a55 g_main_context_iteration (libglib-2.0.so.0 + 0x5ca55)
                #3  0x000077da0a469ab2 n/a (libglib-2.0.so.0 + 0x5cab2)
                #4  0x000077da0a49e026 n/a (libglib-2.0.so.0 + 0x91026)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70957:
                #0  0x000077da0c91abb0 ppoll (libc.so.6 + 0x10bbb0)
                #1  0x000077da0a4cd227 n/a (libglib-2.0.so.0 + 0xc0227)
                #2  0x000077da0a469a55 g_main_context_iteration (libglib-2.0.so.0 + 0x5ca55)
                #3  0x000077d9f836c2fe n/a (libdconfsettings.so + 0x62fe)
                #4  0x000077da0a49e026 n/a (libglib-2.0.so.0 + 0x91026)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70960:
                #0  0x000077da0c91a63d __poll (libc.so.6 + 0x10b63d)
                #1  0x000077d9f95575d7 n/a (libQt6WaylandClient.so.6 + 0x605d7)
                #2  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #3  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #4  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70955:
                #0  0x000077da0c9261fd syscall (libc.so.6 + 0x1171fd)
                #1  0x000077da0a49cc37 g_cond_wait_until (libglib-2.0.so.0 + 0x8fc37)
                #2  0x000077da0a432955 n/a (libglib-2.0.so.0 + 0x25955)
                #3  0x000077da0a432ac8 g_async_queue_timeout_pop (libglib-2.0.so.0 + 0x25ac8)
                #4  0x000077da0a4a2a30 n/a (libglib-2.0.so.0 + 0x95a30)
                #5  0x000077da0a49e026 n/a (libglib-2.0.so.0 + 0x91026)
                #6  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #7  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70959:
                #0  0x000077da0c9261fd syscall (libc.so.6 + 0x1171fd)
                #1  0x000077da0a49bf20 g_cond_wait (libglib-2.0.so.0 + 0x8ef20)
                #2  0x000077da0a43298c n/a (libglib-2.0.so.0 + 0x2598c)
                #3  0x000077da0a4329fd g_async_queue_pop (libglib-2.0.so.0 + 0x259fd)
                #4  0x000077d9faa4b5fc n/a (libpangoft2-1.0.so.0 + 0xc5fc)
                #5  0x000077da0a49e026 n/a (libglib-2.0.so.0 + 0x91026)
                #6  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #7  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 70961:
                #0  0x000077da0c91a63d __poll (libc.so.6 + 0x10b63d)
                #1  0x000077d9f95575d7 n/a (libQt6WaylandClient.so.6 + 0x605d7)
                #2  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #3  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #4  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 71010:
                #0  0x000077da0c89fa19 n/a (libc.so.6 + 0x90a19)
                #1  0x000077da0c8a2479 pthread_cond_wait (libc.so.6 + 0x93479)
                #2  0x000077d9ea45257e n/a (libvulkan_radeon.so + 0x25257e)
                #3  0x000077d9ea42a1bc n/a (libvulkan_radeon.so + 0x22a1bc)
                #4  0x000077d9ea4524dd n/a (libvulkan_radeon.so + 0x2524dd)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 71015:
                #0  0x000077da0c89fa19 n/a (libc.so.6 + 0x90a19)
                #1  0x000077da0c8a27e2 pthread_cond_timedwait (libc.so.6 + 0x937e2)
                #2  0x000077da0d0de3c8 _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x2de3c8)
                #3  0x000077da0d0dfad5 n/a (libQt6Core.so.6 + 0x2dfad5)
                #4  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 71017:
                #0  0x000077da0c89fa19 n/a (libc.so.6 + 0x90a19)
                #1  0x000077da0c8a27e2 pthread_cond_timedwait (libc.so.6 + 0x937e2)
                #2  0x000077da0d0de3c8 _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x2de3c8)
                #3  0x000077da0d0dfad5 n/a (libQt6Core.so.6 + 0x2dfad5)
                #4  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 71018:
                #0  0x000077da0c89fa19 n/a (libc.so.6 + 0x90a19)
                #1  0x000077da0c8a27e2 pthread_cond_timedwait (libc.so.6 + 0x937e2)
                #2  0x000077da0d0de3c8 _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x2de3c8)
                #3  0x000077da0d0dfad5 n/a (libQt6Core.so.6 + 0x2dfad5)
                #4  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 71019:
                #0  0x000077da0c89fa19 n/a (libc.so.6 + 0x90a19)
                #1  0x000077da0c8a27e2 pthread_cond_timedwait (libc.so.6 + 0x937e2)
                #2  0x000077da0d0de3c8 _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x2de3c8)
                #3  0x000077da0d0dfad5 n/a (libQt6Core.so.6 + 0x2dfad5)
                #4  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 71014:
                #0  0x000077da0c89fa19 n/a (libc.so.6 + 0x90a19)
                #1  0x000077da0c8a27e2 pthread_cond_timedwait (libc.so.6 + 0x937e2)
                #2  0x000077da0d0de3c8 _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x2de3c8)
                #3  0x000077da0d0dfad5 n/a (libQt6Core.so.6 + 0x2dfad5)
                #4  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                
                Stack trace of thread 71013:
                #0  0x000077da0c89fa19 n/a (libc.so.6 + 0x90a19)
                #1  0x000077da0c8a27e2 pthread_cond_timedwait (libc.so.6 + 0x937e2)
                #2  0x000077da0d0de3c8 _ZN14QWaitCondition4waitEP6QMutex14QDeadlineTimer (libQt6Core.so.6 + 0x2de3c8)
                #3  0x000077da0d0dfad5 n/a (libQt6Core.so.6 + 0x2dfad5)
                #4  0x000077da0d0d90cd n/a (libQt6Core.so.6 + 0x2d90cd)
                #5  0x000077da0c8a339d n/a (libc.so.6 + 0x9439d)
                #6  0x000077da0c92849c n/a (libc.so.6 + 0x11949c)
                ELF object binary architecture: AMD x86-64

```

</details>

It's likely the app attempts to remove an entry of a folder already deleted. Maybe it could be resolved checking if the entry exists before deleting.

**Comment by zaps166** [comment]
> CANC

You mean, the "DEL" key?

---

I have about 11k entries with subfolders. I did Ctrl+A, Del - the list is empty, no crash.

---

Are you able to compile in Debug and check backtrace again? How did you install QMPlay2 (AUR package)?

**Comment by Digitalone1** [comment]
> You mean, the "DEL" key?

Yes, sorry, I have the italian keybord.

> I have about 11k entries with subfolders. I did Ctrl+A, Del - the list is empty, no crash.

Did you try subfolders with more levels?

> Are you able to compile in Debug and check backtrace again? How did you install QMPlay2 (AUR package)?

Yes, AUR package, but I can't compile right now.

I'm trying to reproduce it, but I'm not able anymore.

If you can't reproduce, just close the issue for the moment.

**Comment by zaps166** [comment]
> Did you try subfolders with more levels?

Yes, a lot of levels. Can't reproduce so far. Is playlist fully loaded and there are no background tasks?

Tried to expand many of them and delete all. Also tried to add them again and delete all - can't crash.

> Yes, AUR package, but I can't compile right now.

`qmplay2` package uses LTO (which is known to cause issues in some software at random - I know Mesa is broken with LTO, once very long time ago somebody reported weird QMPlay2 crash with LTO, too - but all is about GCC, not Clang which is used by QMPlay2 there). You can try `qmplay2-git` which doesn't use LTO. If this helps, I'll get rid of LTO in main package.

---

If you decide to compile in Debug, you can set `-DCMAKE_BUILD_TYPE=Debug -DUSE_ASAN=ON -DCMAKE_CXX_COMPILER=clang++` - maybe this'll help debugging!

**Comment by Digitalone1** [comment]
Ok, I close this for the moment. If I get the crash another time, I'll reopen it.

**Comment by zaps166** [comment]
Ok!

**Comment by zaps166** [comment]
Maybe this could fix it: 0a96c99e7afbebc5a7136ef8d055c5b04d0708fc
